verifier/src/identity.rs

```diff
@@ -200,14 +235,175 @@ fn mitigated_advisories_to_advisories(
+    fn verify(&self, evidence: &E) -> VerificationOutput<Self::Value> {
+        let result = self
+            .identity_verifiers
+            .iter()
+            .find_map(|identity_verifier| {
+                let result = match identity_verifier {
+                    TrustedIdentityVerifier::MrEnclave(mr_enclave) => {
+                        mr_enclave.verifier.verify(evidence).is_success()
+                    }
+                    TrustedIdentityVerifier::MrSigner(mr_signer) => {
+                        mr_signer.verifier.verify(evidence).is_success()
+                    }
+                };
+
+                match result.unwrap_u8() {
+                    1 => Some(VerificationOutput::new((), result)),
+                    _ => None,
+                }
+            });
+
+        result.unwrap_or(VerificationOutput::new((), 0.into()))
+    }
+}
+
+/// A verifier for looking at a single identity
+///
+/// This can be a bit confusing with the `TrustedIdentitiesVerifier`. This type is to handle *one*
+/// identity, while the `TrustedIdentitiesVerifier` is for verifying one of *multiple* identities.
+#[derive(Debug, Clone)]
+enum TrustedIdentityVerifier {
+    MrEnclave(TrustedMrEnclaveIdentityVerifier),
+    MrSigner(TrustedMrSignerIdentityVerifier),
+}
+
+impl From<TrustedIdentity> for TrustedIdentityVerifier {
+    fn from(identity: TrustedIdentity) -> Self {
+        match identity {
+            TrustedIdentity::MrEnclave(mr_enclave) => Self::MrEnclave(mr_enclave.into()),
+            TrustedIdentity::MrSigner(mr_signer) => Self::MrSigner(mr_signer.into()),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TrustedMrEnclaveIdentityVerifier {
+    verifier: And<MrEnclaveVerifier, AdvisoriesVerifier>,
+}
+
+impl From<TrustedMrEnclaveIdentity> for TrustedMrEnclaveIdentityVerifier {
+    fn from(identity: TrustedMrEnclaveIdentity) -> Self {
+        let verifier = And::new(
+            MrEnclaveVerifier::new(identity.mr_enclave()),
+            AdvisoriesVerifier::new(identity.advisories()),
+        );
+        Self { verifier }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TrustedMrSignerIdentityVerifier {
+    verifier: And<MrSignerVerifier, AdvisoriesVerifier>,
+}
+
+impl From<TrustedMrSignerIdentity> for TrustedMrSignerIdentityVerifier {
+    fn from(identity: TrustedMrSignerIdentity) -> Self {
+        let verifier = And::new(
+            MrSignerVerifier::new(
+                identity.mr_signer(),
+                identity.isv_product_id(),
+                identity.isv_svn(),
+            ),
+            AdvisoriesVerifier::new(identity.advisories()),
+        );
+        Self { verifier }
+    }
+}
+
 #[cfg(test)]
 mod test {
     use super::*;
 
+    /// An identity to use in testing the different identity verifiers.
+    struct Identity {
+        mr_enclave: MrEnclave,
+        mr_signer: MrSigner,
+        isv_product_id: IsvProductId,
+        isv_svn: IsvSvn,
+        advisories: Advisories,
+    }
```

Comment:
(first part of the comment is about the overall PR not these lines of code)

This is PR missing the `Display` implementation. 
The PR already got a bit big as it is so stopped where it's at.

I think the Display implementation  of `TrustedIdentiesVerifier` will be one of the more important ones as it's most likely that someone doesn't have new enclave Identity when they fail to attest.  I'm thinking we need to display. 

Having said that I'm thinking that in the display of a failure to verify an Identity we may want to display a type similar to this one, where the MRENCLAVE and MRSIGNER values are present.

Normally for the success case we say what was expected, but I'm thinking saying "one" of these Identities was met isn't that useful and instead we should say which one was matched.

Reply:
That all makes sense.